### PR TITLE
SWATCH-3559: Add time-bound filtering to Billable Usage Account ID APIs

### PR DESCRIPTION
**Pull request by wottop**
Jira issue: SWATCH-3559

## Description

Added filter on BillingAccountId queries for time constraints

## Testing
 
Added to unit tests so that time filtering is expressed.

### CODE REVIEW

**Comment by lindseyburnett** [comment]
# Billing Account ID APIs with Date Filtering

## Overview

This test plan verifies that:

* `/subscriptions/billing_account_ids` returns only **active subscriptions** in the current month.  
* `/instances/billing_account_ids` returns only **hosts with `last_seen` in the current month**.

---

## 1\. Environment Setup

```
# Reset local services
podman-compose down
podman-compose up -d

# Apply database migrations
./gradlew liquibaseUpdate

# Start rhsm-subscriptions service (port 8000)
DEV_MODE=true SERVER_PORT=8000 SPRING_PROFILES_ACTIVE=api ./gradlew :bootRun

# Start swatch-contracts service (port 8001, mgmt port 9001)
SERVER_PORT=8001 QUARKUS_MANAGEMENT_PORT=9001 ./gradlew :swatch-contracts:quarkusDev
```

---

## 2\. Data Setup

### 2.1 Offering, SKU Mapping, Account Services

```
INSERT INTO public.offering (sku, product_name, product_family, cores, sockets, hypervisor_cores, hypervisor_sockets, role, sla, usage, description, has_unlimited_usage, derived_sku, metered, special_pricing_flag, level_1, level_2) VALUES ('MW02393', 'OpenShift Online', 'OpenShift Enterprise', null, null, null, null, null, 'Premium', null, 'Red Hat OpenShift Service on AWS Hosted Control Planes (Hourly)', false, null, true, null, 'OpenShift', 'ROSA - RH OpenShift on AWS');

INSERT INTO public.sku_product_tag (sku, product_tag, id) VALUES ('MW02393', 'rosa', '2248acc7-5aa2-41e6-9af6-a7042521f254');

INSERT INTO public.account_services (account_number, service_type, org_id) VALUES ('5910538', 'rosa Instance', '11009103');
```

### 2.2 Subscriptions Table

Includes a mix of active, inactive, future, and expired subscriptions for testing.

```
-- 1. Start date = today one year ago, no end date
INSERT INTO public.subscription (sku, org_id, subscription_id, quantity, start_date, end_date, billing_provider_id, subscription_number, billing_provider, billing_account_id)
VALUES ('MW02393', '11009103', '12345678', 1, '2024-05-07 21:59:41.415492 +00:00', null, 'ns0m6bivujkq;DYrch150QiQM;555653821384', '15554835', 'aws', '489360417680');

-- 2. Start date = sometime this month, end date = one year later
INSERT INTO public.subscription (sku, org_id, subscription_id, quantity, start_date, end_date, billing_provider_id, subscription_number, billing_provider, billing_account_id)
VALUES ('MW02393', '11009103', 22345678, 1, '2025-05-10 21:59:41.415492 +00:00', '2026-05-10 21:59:41.415492 +00:00', 'yo9flujc3mkh;aWFVel4aQk2J;560487559022', '15608087', 'aws', '705395152839');

-- 3. Start date = last month, end date = 6 months from then
INSERT INTO public.subscription (sku, org_id, subscription_id, quantity, start_date, end_date, billing_provider_id, subscription_number, billing_provider, billing_account_id)
VALUES ('MW02393', '11009103', 32345678, 1, '2025-04-15 21:59:41.415492 +00:00', '2025-10-15 21:59:41.415492 +00:00', '55plhro1xcse;EOPLx2yDSfvB;268789227972', '15731351', 'aws', '979335536418');

-- 4. Start date = 4 months ago, end date = 5 months ago (inverted)
INSERT INTO public.subscription (sku, org_id, subscription_id, quantity, start_date, end_date, billing_provider_id, subscription_number, billing_provider, billing_account_id)
VALUES ('MW02393', '11009103', 42345678, 1, '2025-01-07 21:59:41.415492 +00:00', '2024-12-07 21:59:41.415492 +00:00', 'd8lt13b2anae;LKtcBxaLEh8f;544401500732', '15903402', 'aws', '461592032275');

-- 5. Start date = next month, no end date
INSERT INTO public.subscription (sku, org_id, subscription_id, quantity, start_date, end_date, billing_provider_id, subscription_number, billing_provider, billing_account_id)
VALUES ('MW02393', '11009103', 52345678, 1, '2025-06-01 21:59:41.415492 +00:00', null, 'i3siz7kevfuj;ZOulnRoYYblu;835790553051', '15464624', 'aws', '204620448387');
```

| billing\_account\_id | start\_date | end\_date | Included? | Reason |
| :---- | :---- | :---- | :---- | :---- |
| 489360417680 | 2024-05-07 | null | ✅ | Active all year |
| 705395152839 | 2025-05-10 | 2026-05-10 | ✅ | Starts this month |
| 979335536418 | 2025-04-15 | 2025-10-15 | ✅ | Overlaps with May |
| 461592032275 | 2025-01-07 | 2024-12-07 | ❌ | Ended last year |
| 204620448387 | 2025-06-01 | null | ❌ | Starts next month |

### 2.3 Hosts and Tally Buckets

```
-- 1. Two records with last_seen this month
INSERT INTO public.hosts (id, inventory_id, insights_id, org_id, display_name, subscription_manager_id, is_guest, hypervisor_uuid, hardware_type, num_of_guests, last_seen, is_unmapped_guest, is_hypervisor, cloud_provider, instance_type, instance_id, billing_provider, billing_account_id, last_applied_event_record_date)
VALUES ('50d5a2c5-1dea-49b0-84bd-0086360e5d7a', null, null, '11009103', 'host-this-month-897', null, false, null, null, null, '2025-05-05 12:00:00.000000 +00:00', false, false, null, 'rosa Instance', '6b905580-3105-4996-a5c3-13241ea86edd', 'aws', '730842100581', '2025-05-05 13:00:00.000000 +00:00');
INSERT INTO public.host_tally_buckets (host_id, usage, product_id, sla, as_hypervisor, cores, sockets, measurement_type, billing_provider, billing_account_id, version)
VALUES ('50d5a2c5-1dea-49b0-84bd-0086360e5d7a', 'Production', 'rosa', 'Premium', false, 8, null, 'PHYSICAL', 'aws', '730842100581', 4);

INSERT INTO public.hosts (id, inventory_id, insights_id, org_id, display_name, subscription_manager_id, is_guest, hypervisor_uuid, hardware_type, num_of_guests, last_seen, is_unmapped_guest, is_hypervisor, cloud_provider, instance_type, instance_id, billing_provider, billing_account_id, last_applied_event_record_date)
VALUES ('79cc4765-fbc5-4f97-a073-ff836e0b7fb4', null, null, '11009103', 'host-this-month-382', null, false, null, null, null, '2025-05-05 12:00:00.000000 +00:00', false, false, null, 'rosa Instance', '435b3b8b-f631-4818-a28d-a5ea21b0bb95', 'aws', '615144211475', '2025-05-05 13:00:00.000000 +00:00');
INSERT INTO public.host_tally_buckets (host_id, usage, product_id, sla, as_hypervisor, cores, sockets, measurement_type, billing_provider, billing_account_id, version)
VALUES ('79cc4765-fbc5-4f97-a073-ff836e0b7fb4', 'Production', 'rosa', 'Premium', false, 4, null, 'PHYSICAL', 'aws', '615144211475', 4);

-- 2. Two records with last_seen two months ago
INSERT INTO public.hosts (id, inventory_id, insights_id, org_id, display_name, subscription_manager_id, is_guest, hypervisor_uuid, hardware_type, num_of_guests, last_seen, is_unmapped_guest, is_hypervisor, cloud_provider, instance_type, instance_id, billing_provider, billing_account_id, last_applied_event_record_date)
VALUES ('5a0ec5b1-08dc-475d-91e1-2b3eb08cf789', null, null, '11009103', 'host-two-months-ago-523', null, false, null, null, null, '2025-03-20 10:00:50.124774 +00:00', false, false, null, 'rosa Instance', 'b48f565c-37be-4d23-9b2d-45c46709b387', 'aws', '608323391244', '2025-03-20 11:00:50.124774 +00:00');
INSERT INTO public.host_tally_buckets (host_id, usage, product_id, sla, as_hypervisor, cores, sockets, measurement_type, billing_provider, billing_account_id, version)
VALUES ('5a0ec5b1-08dc-475d-91e1-2b3eb08cf789', 'Production', 'rosa', 'Premium', false, 16, null, 'PHYSICAL', 'aws', '608323391244', 3);

INSERT INTO public.hosts (id, inventory_id, insights_id, org_id, display_name, subscription_manager_id, is_guest, hypervisor_uuid, hardware_type, num_of_guests, last_seen, is_unmapped_guest, is_hypervisor, cloud_provider, instance_type, instance_id, billing_provider, billing_account_id, last_applied_event_record_date)
VALUES ('ecf8804b-bf1b-4843-b7f7-899f06b08a60', null, null, '11009103', 'host-two-months-ago-546', null, false, null, null, null, '2025-03-20 10:00:50.124774 +00:00', false, false, null, 'rosa Instance', 'ed358c75-0e2a-48e9-99c9-a1f05f85c541', 'aws', '580673758477', '2025-03-20 11:00:50.124774 +00:00');
INSERT INTO public.host_tally_buckets (host_id, usage, product_id, sla, as_hypervisor, cores, sockets, measurement_type, billing_provider, billing_account_id, version)
VALUES ('ecf8804b-bf1b-4843-b7f7-899f06b08a60', 'Production', 'rosa', 'Premium', false, 2, null, 'PHYSICAL', 'aws', '580673758477', 2);

-- 3. One record with last_seen last month
INSERT INTO public.hosts (id, inventory_id, insights_id, org_id, display_name, subscription_manager_id, is_guest, hypervisor_uuid, hardware_type, num_of_guests, last_seen, is_unmapped_guest, is_hypervisor, cloud_provider, instance_type, instance_id, billing_provider, billing_account_id, last_applied_event_record_date)
VALUES ('b6763c3a-981d-4d90-8609-1c6dadf82ec2', null, null, '11009103', 'host-last-month-756', null, false, null, null, null, '2025-04-15 14:00:50.124774 +00:00', false, false, null, 'rosa Instance', '97ecbc51-c449-43d7-aa60-197a1dcbe17c', 'aws', '349865472335', '2025-04-15 15:00:50.124774 +00:00');
INSERT INTO public.host_tally_buckets (host_id, usage, product_id, sla, as_hypervisor, cores, sockets, measurement_type, billing_provider, billing_account_id, version)
VALUES ('b6763c3a-981d-4d90-8609-1c6dadf82ec2', 'Production', 'rosa', 'Premium', false, 1, null, 'PHYSICAL', 'aws', '349865472335', 5);

```

| billing\_account\_id | last\_seen | Included? | Reason |
| :---- | :---- | :---- | :---- |
| 730842100581 | 2025-05-05 | ✅ | Seen this month |
| 615144211475 | 2025-05-05 | ✅ | Seen this month |
| 608323391244 | 2025-03-20 | ❌ | Two months ago |
| 580673758477 | 2025-03-20 | ❌ | Two months ago |
| 349865472335 | 2025-04-15 | ❌ | Last month |

---

## 3\. API Tests

### 3.1 `/subscriptions/billing_account_ids`

**Command:**

```
http GET :8001/api/swatch-contracts/v1/subscriptions/billing_account_ids \
  product_tag==rosa \
  org_id==11009103
```

**Expected Output:**

```
{
  "ids": [
    {
      "org_id": "11009103",
      "product_tag": "rosa",
      "billing_provider": "aws",
      "billing_account_id": "489360417680"
    },
    {
      "org_id": "11009103",
      "product_tag": "rosa",
      "billing_provider": "aws",
      "billing_account_id": "705395152839"
    },
    {
      "org_id": "11009103",
      "product_tag": "rosa",
      "billing_provider": "aws",
      "billing_account_id": "979335536418"
    }
  ]
}

```

---

### 3.2 `/instances/billing_account_ids`

**Command:**

```
http GET :8000/api/rhsm-subscriptions/v1/instances/billing_account_ids \
  product_tag==rosa \
  org_id==11009103 \
  x-rh-identity:$(echo -n '{"identity":{"internal":{"org_id":"11009103"}}}' | base64 -w0)
```

**Expected Output:**

```
{
  "ids": [
    {
      "org_id": "11009103",
      "product_tag": "rosa",
      "billing_provider": "aws",
      "billing_account_id": "615144211475"
    },
    {
      "org_id": "11009103",
      "product_tag": "rosa",
      "billing_provider": "aws",
      "billing_account_id": "730842100581"
    }
  ]
}

```

File: swatch-contracts/src/main/java/com/redhat/swatch/contract/repository/SubscriptionRepository.java [hunk 1]
```diff
@@ -90,7 +90,12 @@ public List<BillingAccountInfoDTO> findBillingAccountInfo(
       String orgId, Optional<String> productTag) {
     StringBuilder query =
         new StringBuilder(
-            "select distinct subscription.orgId, subscription.billingAccountId, subscription.billingProvider, productTag from SubscriptionEntity subscription join subscription.offering offering join offering.productTags productTag where orgId = ?1");
+            "select distinct subscription.orgId, subscription.billingAccountId, subscription.billingProvider, productTag from SubscriptionEntity subscription "
+                + "join subscription.offering offering "
+                + "join offering.productTags productTag "
+                + "where orgId = ?1 "
+                + "and subscription.startDate <= CURRENT_TIMESTAMP "
+                + "and subscription.endDate >= CURRENT_TIMESTAMP");
```
Comment (lindseyburnett) [inline swatch-contracts/src/main/java/com/redhat/swatch/contract/repository/SubscriptionRepository.java:98]: Empty end dates need to be taken into consideration for this.  See the comment with the test plan/expected results.

If a subscription started on May 7, 2025 and doesn't have an end date, I would expect its billing account id to be part of the results.

**Review by lindseyburnett — CHANGES_REQUESTED** [review]

**Comment by InsightsDroid** [comment]
### IQE Tests Summary Report
 10 failed, 234 passed, 2 skipped, 242 deselected, 310 warnings in 2522.13s (0:42:02) 
Result: Failed
IQE plugin image: https://quay.io/cloudservices/iqe-tests:rhsm-subscriptions
|Failures|
|:-| 
|integration/exports/test_export_instances.py::test_export_instances_metric_filters[json-red| 
|integration/exports/test_export_instances.py::test_export_instances_metric_filters[csv-red| 
|integration/swatch_billing/test_remitance.py::test_verify_remittance_of_metered_product[red| 
|integration/swatch_billing/test_remitance.py::test_verify_end_to_end_payg_metrics| 
|integration/swatch_billing/test_remitance.py::test_verify_remittance_excludes_hard_failures[OpenShift-metrics-aws]| 
|integration/swatch_billing/test_remitance.py::test_verify_remittance_of_contract_product[rosa-greater]| 
|integration/swatch_billing/test_remitance.py::test_verify_rosa_vs_osd_in_swatch| 
|integration/swatch_billing/test_remitance.py::test_verify_payg_usage_when_billing_account_id_updated[rhel-for-x86-els-payg]| 
|integration/swatch_tally/test_tally_dedup_rhel_els.py::test_verify_tally_sync_different_event_source[quantities0]| 
|integration/swatch_tally/test_tally_dedup_rhel_els.py::test_verify_tally_sync_different_event_source[quantities1]|

**Comment by kartikshahc** [comment]
For subscriptions table
 
| billing\_account\_id | start\_date | end\_date | Included? | Reason |Test coverage
| :---- | :---- | :---- | :---- | :---- | :----
| 489360417680 | 2024-05-07 | null | ✅ | Started last year, still active | Yes |
| 979335536418 | 2025-04-15 | 2025-10-15 | ✅ | Active as of May 7 | Yes |
| 461592032275 | 2024-01-07 | 2024-12-07 | ❌ | Ended before today | Yes |
| 705395152839 | 2025-05-10 | 2026-05-10 | ❌ | Future-dated start (not active yet) | Yes |
| 204620448387 | 2025-06-01 | null | ❌ | Starts next month | Yes |

File: swatch-contracts/src/main/java/com/redhat/swatch/contract/repository/SubscriptionRepository.java (hunk 1 above)
Comment (kartikshahc) [reply swatch-contracts/src/main/java/com/redhat/swatch/contract/repository/SubscriptionRepository.java]: Done

**Comment by InsightsDroid** [comment]
### IQE Tests Summary Report
 243 passed, 2 skipped, 240 deselected, 309 warnings in 2015.58s (0:33:35) 
Result: Success
IQE plugin image: https://quay.io/cloudservices/iqe-tests:rhsm-subscriptions

**Comment by InsightsDroid** [comment]
### IQE Tests Summary Report
 243 passed, 2 skipped, 240 deselected, 309 warnings in 1974.19s (0:32:54) 
Result: Success
IQE plugin image: https://quay.io/cloudservices/iqe-tests:rhsm-subscriptions

**Review by lindseyburnett — APPROVED** [review]

**Comment by InsightsDroid** [comment]
### IQE Tests Summary Report
 243 passed, 2 skipped, 240 deselected, 309 warnings in 2139.17s (0:35:39) 
Result: Success
IQE plugin image: https://quay.io/cloudservices/iqe-tests:rhsm-subscriptions

**Review by TrayvonMcKnight — COMMENTED** [review]
This looks good to me.

We already have a few tests that are hitting the subscriptions/billing_account_id endpoints 
`test_get_billing_account_ids_no_product_tag`
`test_get_billing_account_ids`

I am adding a test to do check for the [empty list/boundary](https://gitlab.cee.redhat.com/insights-qe/iqe-rhsm-subscriptions-plugin/-/merge_requests/1168)

**Review by TrayvonMcKnight — APPROVED** [review]

**Comment by InsightsDroid** [comment]
### IQE Tests Summary Report
 243 passed, 2 skipped, 240 deselected, 309 warnings in 1946.11s (0:32:26) 
Result: Success
IQE plugin image: https://quay.io/cloudservices/iqe-tests:rhsm-subscriptions
